### PR TITLE
Basic orchestration layer for python-observabilityclient

### PR DESCRIPTION
**Pull request by paramite**
This patch adds following features:
  - container orchestration role
  - thanos deployment playbook for testing purposes
  - prometheus agent deployment playbook for observability client

Co-authored-by: Marihan Girgis <mgirgisf@redhat.com>
Partially-Implements: OSP-14664
Related: https://github.com/infrawatch/python-observabilityclient/pull/1

### CODE REVIEW

File: playbooks/prometheus_agent.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+
+- name: Deploy Prometheus agent
+  hosts: {! prometheus_destination !}
+  become: true
+  become_methos: sudo
```
Comment (csibbitt) [inline playbooks/prometheus_agent.yml.j2]: become_methoD

File: playbooks/prometheus_agent.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,66 @@
+---
+
+- name: Deploy Prometheus agent
+  hosts: {! prometheus_destination !}
```
Comment (csibbitt) [inline playbooks/prometheus_agent.yml.j2]: What do the `{!...!}` tags mean? I'm not familiar with that syntax and having trouble finding reference to it in jinja/ansible docs

Aha! https://github.com/infrawatch/python-observabilityclient/pull/1/files#diff-66243527cb640bca2d1f4a8a2852a74743e34e09712ad0e8f53cd54b48875397R35  

I'm wondering if there is a less complicated way to do this. I'll keep reading.

File: plugins/modules/puppet_environment.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            - List of variables to load
+        required: true
+
+author:
+    - Martin Mágr (mmagr@redhat.com)
+'''
+
+EXAMPLES = '''
+- name: Test module
+  puppet_environment:
```
Comment (csibbitt) [inline plugins/modules/puppet_environment.py:54]: Is this being used anywhere? I am not spotting it.

File: roles/spawn_container/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    path: "{{ item }}"
+    state: directory
+    owner: root
+    group: root
+    mode: '0750'
+  with_items:
+    - /var/log/containers
+    - "/var/log/containers/{{ name }}"
+
+- name: Container clenup
```
Comment (csibbitt) [inline roles/spawn_container/tasks/main.yml]: cleanup

File: roles/spawn_container/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
+---
```
Comment (csibbitt) [inline roles/spawn_container/tasks/main.yml:1]: Are there other openstackclient plugins that end up managing containers like this? I'm wondering if there is a potential to re-use the same code for this sort of thing.

File: playbooks/prometheus_agent.yml.j2 (hunk 2 above)
Comment (csibbitt) [reply playbooks/prometheus_agent.yml.j2]: We had an out-of-band discussion on this, and I think that the templating of this playbook can be eliminated:

1. Pass `workdir` `prometheus_container_name`, and `prometheus_destination` as "extravars" here: https://github.com/infrawatch/python-observabilityclient/pull/1/files#diff-a1b0ac567ec121f9438fe42cf7408aecb8c765251b1b666870c70065cb35ffb1R158
2. Set sane defaults for these. It looks like workdir is done, we can simply choose the prometheus_container_name we want (my fist suggestion: osp_obs_prometheus_agent), and I think prometheus_destination could default to "allovercloud"

File: roles/spawn_container/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    generate_systemd:
+      container_prefix: "osp-observability"
+      path: /etc/systemd/system/
+      restart_policy: always
+  register: result_no_volumes
+  when: volumes is not defined or (volumes is defined and (volumes|length<0))
+
+- name: Get result
+  set_fact:
+    result: "{% if result_with_volumes.container is defined %}{{ result_with_volumes }}{% else %}{{ result_no_volumes }}{% endif %}"
```
Comment (csibbitt) [inline roles/spawn_container/tasks/main.yml:57]: Pretty sure you could just register "result"  directly (in both cases) and not bother with this logic

File: roles/spawn_container/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
+---
+
+- name: Check if mount points exist
+  include_tasks: volume_check.yml
+  with_items:
+    - "{{ volumes }}"
+  when: volumes is defined and (volumes|length>0)
```
Comment (csibbitt) [inline roles/spawn_container/tasks/main.yml]: You can eliminate all of these `volumes is defined and ...` clauses by setting an empty default for volumes and just testing the length

File: roles/spawn_container/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    state: started
+    restart_policy: "always"
+    network: host
+    command: "{{ command }}"
+    generate_systemd:
+      container_prefix: "osp-observability"
+      path: /etc/systemd/system/
+      restart_policy: always
+  register: result_no_volumes
+  when: volumes is not defined or (volumes is defined and (volumes|length<0))
```
Comment (csibbitt) [inline roles/spawn_container/tasks/main.yml]: I think this should be: volumes|length **==** 0

**Review by csibbitt — CHANGES_REQUESTED** [review]
I think the templating of the playbook should be eliminated before we move on from this starting point.

File: plugins/modules/puppet_environment.py (hunk 3 above)
Comment (paramite) [reply plugins/modules/puppet_environment.py:54]: It is not yet used, but I suspect we gonna need to get OpenStack configuration knowledge sooner or later. This is a resulting code from a initial spike about getting tripleo configuration values

File: roles/spawn_container/tasks/main.yml (hunk 5 above)
Comment (paramite) [reply roles/spawn_container/tasks/main.yml:1]: There is tripleoclient plugin that has very similar logic, but tightly coupled with tripleo itself. Which is a reason why we don't have our commands as part of tripleoclient and have our own plugin instead.

File: plugins/modules/puppet_environment.py (hunk 3 above)
Comment (csibbitt) [reply plugins/modules/puppet_environment.py:54]: In a way, I hope you are wrong and we do not need any other OSP configuration after we have the inventory of hosts. OTOH, this is useful code, so thanks for publishing it!

File: roles/spawn_container/tasks/main.yml (hunk 6 above)
Comment (paramite) [reply roles/spawn_container/tasks/main.yml:57]: Well and yet it didn't work, when I used "result" for both execution branches :)

Comment (csibbitt) [reply roles/spawn_container/tasks/main.yml:57]: Interesting! When I get this running I'll try it myself and maybe I'll learn something.

File: playbooks/prometheus_agent.yml [hunk 9]
```diff
@@ -0,0 +1,71 @@
+---
+
+- name: Deploy Prometheus agent
+  hosts: "{{ prometheus_destination | default('controller-0') }}"
```
Comment (csibbitt) [inline playbooks/prometheus_agent.yml:4]: suggest `default('allovercloud')`

**Comment by csibbitt** [comment]
Reviewed and tested this again today. I'm going to Approve it now because it appears to work, but there are still typos and potentially a logic error with volumes|length that have not been resolved since my last review; so please consider making the changes before merging.

**Review by csibbitt — APPROVED** [review]

**Comment by csibbitt** [comment]
Changes LGTM, thanks!  Approval stands. :)

File: playbooks/prometheus_agent.yml [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ---
 
 - name: Deploy Prometheus agent
-  hosts: "{{ prometheus_destination | default('controller-0') }}"
+  hosts: "{{ prometheus_destination | default('allovercloud') }}"
```
Comment (csibbitt) [inline playbooks/prometheus_agent.yml]: Sorry, I think this was right the first time. I somehow had a picture in my head of an agent on each node that scrapes the local containers, but that is really not what we discussed. My apologies for the confusion.

 https://github.com/infrawatch/python-observabilityclient/pull/1#issuecomment-1184624887

Comment (csibbitt) [reply playbooks/prometheus_agent.yml]:
```suggestion
  hosts: "{{ prometheus_destination | default('controller-0') }}"
```

**Review by csibbitt — APPROVED** [review]

**Pull request merged**